sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,500 @@
+      session = cluster.connect();
+      resultSet = session.execute(spec.getQuery());
+      final MappingManager mappingManager = new MappingManager(session);
+      Mapper mapper = (Mapper) mappingManager.mapper(spec.getEntityName());
+      iterator = mapper.map(resultSet).iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public void close() {
+      LOG.debug("Closing Cassandra reader");
+      session.close();
+      cluster.close();
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public T getCurrent() {
+      return current;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      return Instant.now();
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write into Apache Cassandra. See {@link CassandraIO} for more
+   * information on usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+
+    abstract Builder<T> builder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setConnectionConfiguration(ConnectionConfiguration
+                                                         connectionConfiguration);
+      abstract Write<T> build();
+    }
+
+    public Write<T> withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "CassandraIO.write()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      input.apply(ParDo.of(new WriteFn<T>(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+  }
+
+  private static class WriteFn<T> extends DoFn<T, Void> {
+
+    private final Write<T> spec;
+
+    private transient Cluster cluster;
+    private transient Session session;
+    private transient MappingManager mappingManager;
+
+    public WriteFn(Write<T> spec) {
+      this.spec = spec;
+    }
+
+    @Setup
+    public void setup() throws Exception {
+      LOG.debug("Starting Cassandra writer");
+      cluster = spec.getConnectionConfiguration().getCluster();
+      session = cluster.connect(spec.getConnectionConfiguration().getKeyspace());
+      mappingManager = new MappingManager(session);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) throws Exception {
+      T entity = context.element();
+      Mapper<T> mapper = (Mapper<T>) mappingManager.mapper(entity.getClass());
+      mapper.save(entity);
```

Comment:
I understand why you're not using batching - because it's faster not to use it - but please link to an explanation of that.
Also: is this synchronous? If not, you need to wait for it to complete, and throw any exceptions in FinishBundle.